### DISCUSSION TITLE
how to set systemPrompt in code?

### DISCUSSION THREAD

**Discussion by eisneim** [withcatai/node-llama-cpp#67] [Q&A] [answered]
the documentation an wraper is very hard to understand, lack of examples: https://withcatai.github.io/node-llama-cpp/guide/chat-prompt-wrapper

for example wizardCoder: 
```
Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
{prompt}

### Response:
```
how to extend this class? the example didn't show how to set systemPrompt
```
class MyCustomChatPromptWrapper extends ChatPromptWrapper {
    public readonly wrapperName: string = "MyCustomChat";
    
    public override wrapPrompt(prompt: string, {systemPrompt, promptIndex}: {systemPrompt: string, promptIndex: number}) {
        if (promptIndex === 0) {
            return "SYSTEM: " + systemPrompt + "\nUSER: " + prompt + "\nASSISTANT:";
        } else {
            return "USER: " + prompt + "\nASSISTANT:";
        }
    }

    public override getStopStrings(): string[] {
        return ["USER:"];
    }

    public override getDefaultStopString(): string {
        return "USER:";
    }
}
```

and also how to write wraper for codefusion
```
<|role_start|>system<|role_end|>{system_message}
<|role_start|>human<|role_end|>{prompt}
<|role_start|>bot<|role_end|>

```

**Comment by giladgd** [comment, marked as answer]
You can pass a custom [`systemPrompt`](https://withcatai.github.io/node-llama-cpp/api/type-aliases/LlamaChatSessionOptions#systemprompt) when you create a [`LlamaChatSession`](https://withcatai.github.io/node-llama-cpp/api/classes/LlamaChatSession) object.

Regarding your `wizardCoder` example, you can just customize the [`GeneralChatPromptWrapper`](https://withcatai.github.io/node-llama-cpp/api/classes/GeneralChatPromptWrapper) with the role names you mentioned:
```typescript
import {fileURLToPath} from "url";
import path from "path";
import {
    LlamaModel, LlamaContext, LlamaChatSession, GeneralChatPromptWrapper
} from "node-llama-cpp";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const model = new LlamaModel({
    modelPath: path.join(__dirname, "models", "codellama-13b.Q3_K_M.gguf"),
    promptWrapper: new GeneralChatPromptWrapper({
        instructionName: "Instruction",
        responseName: "Response"
    })
});
const context = new LlamaContext({model});
const session = new LlamaChatSession({
    context,
    systemPrompt: "Below is an instruction that describes a task. Write a response that appropriately completes the request."
});


const q1 = "Hi there, how are you?";
console.log("User: " + q1);

const a1 = await session.prompt(q1);
console.log("AI: " + a1);


const q2 = "Summerize what you said";
console.log("User: " + q2);

const a2 = await session.prompt(q2);
console.log("AI: " + a2);
```

Based on this example, you can read the source code of [`GeneralChatPromptWrapper.ts`](https://github.com/withcatai/node-llama-cpp/blob/master/src/chatWrappers/GeneralChatPromptWrapper.ts) for an example of how it was implemented, so you can create something similar for `codefusion`.

You're also welcome to open a PR with a `CodeFusionChatPromptWrapper` so other users can just use it out of the box. You can read the [development guide](https://withcatai.github.io/node-llama-cpp/guide/development) for instructions on how to do that.

**Comment by eisneim** [comment]
thank you for the answer!
